File: src/main/java/photos/brooklyn/practice/codingexercises/java/leetcode/PrimePalindrom.java
package photos.brooklyn.practice.codingexercises.java.leetcode;

/**
 * https://leetcode.com/contest/weekly-contest-92/problems/prime-palindrome/
 */
public class PrimePalindrom {
    public int primePalindrome(int N) {
        if(N<=11) return nextPrime(N);
        return (int)findPrime(closestPalindrom(N));
    }

    private static int nextPrime(int n) {
        return isPrime(n) ? n : nextPrime(n+1);
    }

    public static long closestPalindrom(int n) {
        return isPalindrom(n) ? n : nextPalindrom(n);
    }

    public static boolean isPalindrom(int n) {
        final String str = Integer.toString(n);
        for(int i=0;i<str.length()/2;i++){
            if(str.charAt(i)!=str.charAt(str.length()-i-1)) return false;
        }
        return true;
    }

    public long findPrime(long testPalindrom) {
        if(isPrime(testPalindrom)) return testPalindrom;
        return findPrime(nextPalindrom(testPalindrom));
    }

    public static long nextPalindrom(long palindrom) {
        final String str = Long.toString(palindrom);
        if(str.length()%2==0){
            final String half = str.substring(0,(str.length()/2));
            final long fromHalf = makePalindrom(half,str.length());
            if(fromHalf>palindrom) return fromHalf;
            final String newHalf = Long.toString(Long.parseLong(half)+1);
            if(half.length()==newHalf.length()){
                return Long.parseLong(newHalf+reverse(newHalf));
            }else{
                return Long.parseLong(newHalf+reverse(newHalf).substring(1));
            }
        }else{
            final String half = str.substring(0,(str.length()/2)+1);
            final long fromHalf = makePalindrom(half,str.length());
            if(fromHalf>palindrom) return fromHalf;
            final String newHalf = Long.toString(Long.parseLong(half)+1);
            if(half.length()==newHalf.length()){
                return Long.parseLong(newHalf+reverse(newHalf).substring(1));
            }else{
                return Long.parseLong(newHalf+reverse(newHalf).substring(1));
            }
        }
    }

    public static long makePalindrom(String half, int length) {
        if(length%2==0){
            return Long.parseLong(half+reverse(half));
        }else{
            return Long.parseLong(half+reverse(half).substring(1));
        }
    }

    public static String reverse(String str) {
        final char[] r = new char[str.length()];
        for(int i=0;i<str.length();i++){
            r[i] = str.charAt(str.length()-i-1);
        }
        return new String(r);
    }

    private static boolean isPrime(long x) {
        if(x<=1) return false;
        if(x<=3) return true;
        if(x%2==0 || x%3==0) return false;
        for(long i=5;(i*i)<=x;i+=6){
            if(x % i ==0 || x % (i+2)==0) return false;
        }
        return true;
    }
}

File: src/main/scala/photos/brooklyn/practice/codingexercises/scala/leetcode/PrimePalindrom.scala
package photos.brooklyn.practice.codingexercises.scala.leetcode

import scala.annotation.tailrec

/**
  * based on solution here: https://leetcode.com/problems/prime-palindrome/discuss/146798/Search-Palindrome-with-Even-Digits
  */
object PrimePalindrom {

  /**
    * creates a palindrom number of odd digits; all even digit palindoms are divisible by 11 so only 11 is prime
    * @param palindromIndex
    * @return
    */
  def makeOddDigitPalindrom(palindromIndex: Int): Int = palindromIndex.toString+palindromIndex.toString.reverse.substring(1) toInt

  @tailrec
  def isPrime(x: Int, tester: Int=3): Boolean =
    if(x<2 || x%2==0) x==2
    else if(tester*tester > x) true
    else if(x%tester==0) false
    else isPrime(x, tester+2)

  @tailrec
  def findPrimePalindrome(n: Int, palindromIndex: Int): Int = makeOddDigitPalindrom(palindromIndex) match {
    case x if x>=n && isPrime(x) => x
    case _ => findPrimePalindrome(n, palindromIndex+1)
  }

  def primePalindrome(N: Int): Int = if(8<=N && N<=11) 11 else findPrimePalindrome(N, 1)

}
